kvserver: deflake TestPromoteNonVoterInAddVoter

tbg opened the pull request:
It was checking the rangelog for LEARNER additions, however it could accidentally pick up changes from starting the cluster that had no relation to the recent zone config change the test wanted to learn the effects of.

Clear the rangelog before changing the zone config to avoid this.

Also, make the explanation of the test correct - it really threw me off because it suggested voter and non-voter counts that really did not match up at all with the reality of the test.

Fixes #142956.

Epic: none
Release note: none

tbg committed:
kvserver: deflake TestPromoteNonVoterInAddVoter

It was checking the rangelog for LEARNER additions, however
it could accidentally pick up changes from starting the cluster
that had no relation to the recent zone config change the test
wanted to learn the effects of.

Clear the rangelog before changing the zone config to avoid this.

Also, make the explanation of the test correct - it really threw me off because
it suggested voter and non-voter counts that really did not match up at all
with the reality of the test.

Fixes #142956.

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142972)
<!-- Reviewable:end -->

tbg committed:
kvserver: improve upreplication detection

The previous method of counting replicas may have terminated
prematurely, exposing the remainder of the test to unexpected
results.

For example, the replica count condition would be satisfied in the presence of
LEARNERs, but LEARNERs are sometimes rolled back.

kvoli reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142972#-:-OLZDYSy1oVzt1B2OGxN:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @tbg)



<!-- Sent from Reviewable.io -->

tbg commented:
TFTR!

bors r+

tbg commented:
bors r+
